cpp/src/community/louvain_impl.cuh

```diff
@@ -82,11 +84,130 @@ std::pair<std::unique_ptr<Dendrogram<vertex_t>>, weight_t> louvain(
-                          dendrogram->current_level_begin(),
-                          dendrogram->current_level_size(),
-                          current_graph_view.local_vertex_partition_range_first());
+    if (rng_state) {
+      rmm::device_uvector<vertex_t> random_cluster_assignments(
+        current_graph_view.local_vertex_partition_range_size(), handle.get_stream());
+
+      // generate as many number as #local_vertices on each GPU
+      detail::sequence_fill(handle.get_stream(),
+                            random_cluster_assignments.begin(),
+                            random_cluster_assignments.size(),
+                            current_graph_view.local_vertex_partition_range_first());
+
+      // shuffle/permute locally
+      rmm::device_uvector<float> random_numbers(random_cluster_assignments.size(),
+                                                handle.get_stream());
+
+      cugraph::detail::uniform_random_fill(handle.get_stream(),
+                                           random_numbers.data(),
+                                           random_numbers.size(),
+                                           float{0.0},
+                                           float{1.0},
+                                           *rng_state);
+      thrust::sort_by_key(handle.get_thrust_policy(),
+                          random_numbers.begin(),
+                          random_numbers.end(),
+                          random_cluster_assignments.begin());
+
+      if constexpr (multi_gpu) {
+        // distribute shuffled/permuted numbers to other GPUs
+        auto& comm           = handle.get_comms();
+        auto const comm_size = comm.get_size();
+        auto const comm_rank = comm.get_rank();
+
+        std::vector<size_t> tx_value_counts(comm_size);
+        std::fill(tx_value_counts.begin(),
+                  tx_value_counts.end(),
+                  random_cluster_assignments.size() / comm_size);
+
+        std::vector<vertex_t> h_random_gpu_ranks;
+        {
+          rmm::device_uvector<vertex_t> d_random_numbers(
+            random_cluster_assignments.size() % comm_size, handle.get_stream());
+          cugraph::detail::uniform_random_fill(handle.get_stream(),
+                                               d_random_numbers.data(),
+                                               d_random_numbers.size(),
+                                               vertex_t{0},
+                                               vertex_t{comm_size},
+                                               *rng_state);
+
+          h_random_gpu_ranks.resize(d_random_numbers.size());
+
+          raft::update_host(h_random_gpu_ranks.data(),
+                            d_random_numbers.data(),
+                            d_random_numbers.size(),
+                            handle.get_stream());
+        }
+
+        for (int i = 0; i < static_cast<int>(random_cluster_assignments.size() % comm_size); i++) {
```

Comment:
I'm not sure what this loop (perhaps this block of code) is accomplishing.  Suppose we have 100 vertices spread across 2 GPUs where GPU 0 has 52 vertices and GPU 1 has 48 vertices.  comm_size will be 2.

d_random_numbers in this code block is going to be (52 % 2) long (which is 0 elements) on GPU 0, and (48 % 2) long (which is 0 elements) on GPU 1.  

`h_random_gpu_ranks` will also have no elements, so this code block will AFAICT, do nothing.  I haven't thought through what the shuffle_values call below will do... probably overwrite random_cluster_assignments with empty arrays on each node.

Reply:
This code is relevant only when # local vertices % comm_size != 0.

We initially assign # local vertices / comm_size to each rank, then, for the remaining # local vertices % comm_size, this logic decides how to distribute the remaining.

My concern is that assigning # local vertices / comm_size to each rank does not achieve uniform random distribution.

I think we should do something like the following.
```
generate a sequence of vertex IDs [local_vertex_partition_range_first, local_vertex_partition_range_last).
if (multi_gpu) {
  Generate a random number in [0, comm_size) per vertex ID and assign a rank based on this random number.
  shuffle.
}
locally shuffle.
if (multi_gpu) {
  move based on extras/deficits.
}
```

Reply:
Umm... wait a second. Let me think more about this...

Reply:
Yeah... I think the logic above will work.

Reply:
@seunghwak Thank for address @ChuckHastings's question. I was about to do it after refactoring that part of the code into a new utility function `permute_range`.
Your suggestion on generation a random rank for each vertex_id is also implement in `permute_range` function.